src/main/webapp/app/file-upload-assessment/file-upload-assessment.component.ts

```diff
@@ -0,0 +1,482 @@
+    }
+
+    public ngOnInit(): void {
+        this.busy = true;
+
+        // Used to check if the assessor is the current user
+        this.accountService.identity().then(user => {
+            this.userId = user!.id!;
+        });
+        this.isAtLeastInstructor = this.accountService.hasAnyAuthorityDirect(['ROLE_ADMIN', 'ROLE_INSTRUCTOR']);
+
+        this.route.params.subscribe(params => {
+            const exerciseId = Number(params['exerciseId']);
+            const submissionValue = params['submissionId'];
+            const submissionId = Number(submissionValue);
+            if (submissionValue === 'new') {
+                this.loadOptimalSubmission(exerciseId);
+            } else {
+                this.loadSubmission(submissionId);
+            }
+        });
+    }
+
+    attachmentExtension(filePath: string): string {
+        if (!filePath) {
+            return 'N/A';
+        }
+
+        return filePath.split('.').pop()!;
+    }
+
+    private loadOptimalSubmission(exerciseId: number): void {
+        this.fileUploadSubmissionService.getFileUploadSubmissionForExerciseWithoutAssessment(exerciseId, true).subscribe(
+            (submission: FileUploadSubmission) => {
+                this.handleReceivedSubmission(submission);
+                // Update the url with the new id, without reloading the page, to make the history consistent
+                const newUrl = window.location.hash.replace('#', '').replace('new', `${this.submission!.id}`);
+                this.location.go(newUrl);
+            },
+            (error: HttpErrorResponse) => {
+                if (error.status === 404) {
+                    // there is no submission waiting for assessment at the moment
+                    this.goToExerciseDashboard();
+                    this.jhiAlertService.info('artemisApp.tutorExerciseDashboard.noSubmissions');
+                } else if (error.error && error.error.errorKey === 'lockedSubmissionsLimitReached') {
+                    this.goToExerciseDashboard();
+                } else {
+                    this.onError('artemisApp.assessment.messages.loadSubmissionFailed');
+                }
+            },
+        );
+    }
+
+    private loadSubmission(submissionId: number): void {
+        this.fileUploadSubmissionService
+            .get(submissionId)
+            .pipe(filter(res => !!res))
+            .subscribe(
+                res => {
+                    this.handleReceivedSubmission(res.body!);
+                    this.loadFeedbacks(this.result.feedbacks);
+                },
+                (error: HttpErrorResponse) => {
+                    if (error.error && error.error.errorKey === 'lockedSubmissionsLimitReached') {
+                        this.goToExerciseDashboard();
+                    } else {
+                        this.onError('');
+                    }
+                },
+            );
+    }
+
+    private handleReceivedSubmission(submission: FileUploadSubmission): void {
```

Comment:
Very generic method name. What does it mean that the receivedSubmission is handled?
I also wonder why we call `loadFeedbacks` both in `loadSubmissions` and here, even though `loadSubmissions` already calls it :thinking: This has an indication of spaghetti code to me.

Reply:
You are right about `loadFeedbacks`, it should be called only in `handleReceivedSubmission`. I changed the name `handleReceivedSubmission` to `initializePropertiesFromSubmission` to make it more clear that we initialize component properties from submission here